skbio/metadata/_interval.py

```diff
@@ -0,0 +1,699 @@
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import operator
+from functools import wraps
+
+from ._intersection import IntervalTree
+from skbio.util._decorator import experimental
+
+
+class Interval:
+    """Stores the location and metadata of an interval feature.
+
+    This class stores an interval feature. An interval feature
+    is defined as a sub-region of a sequence that is a functional
+    entity, e.g., a gene, a riboswitch, an exon, etc. It can span
+    a single contiguous region or multiple non-contiguous regions (e.g.
+    multiple exons in a transcript, or multiple genes in an operon).
+
+    Parameters
+    ----------
+    interval_metadata : object
+        A reference to the ``IntervalMetadata`` object that this
+        ``Interval`` object is associated to.
+    locations : iterable of tuple of ints
+        List of tuples representing start and end coordinates.
+    boundaries : iterable of tuple of bool
+        List of tuples, representing the openness of each location coordinates.
+        If this isn't specified, then all of the boundaries are True. If any
+        of the coordinate boundaries is False, it indicates that the exact
+        boundary point of a interval feature is unknown. The location may
+        begin or end at some points outside the specified coordinates. This
+        accommodates the location format [1]_ of INSDC.
+    metadata : dict
+        Dictionary of attributes storing information of the feature
+        such as "strand", "gene_name", or "product".
+
+    Attributes
+    ----------
+    locations
+    boundaries
+    metadata
+
+    See Also
+    --------
+    skbio.metadata.IntervalMetadata
+
+    Examples
+    --------
+    >>> from skbio.metadata import Interval, IntervalMetadata
+    >>> gene = Interval(interval_metadata=IntervalMetadata(),
+    ...                 locations=[(1, 2), (4, 7)],
+    ...                 metadata={'name': 'sagA'})
+    >>> gene    # doctest: +ELLIPSIS
+    Interval(interval_metadata=..., locations=[(1, 2), (4, 7)], \
+boundaries=[(True, True), (True, True)], metadata={'name': 'sagA'})
+
+    References
+    ----------
+    .. [1] ftp://ftp.ebi.ac.uk/pub/databases/embl/doc/FT_current.html#3.4.3
+    """
+    def __init__(self, interval_metadata, locations,
+                 boundaries=None, metadata=None):
+        if not isinstance(interval_metadata, IntervalMetadata):
+            raise TypeError('You need to provide an IntervalMetadata'
+                            'object, not %r' % interval_metadata)
+        # Intervals
+        self._interval_metadata = interval_metadata
+
+        locations = list(locations)
+        self.locations = locations
```

Comment:
sorry, the line number was a bit messed up - I mean `list` is needed because of line 80 if the `locations` is an iterator.